internal/rbd/rbd_healer.go

```diff
@@ -0,0 +1,202 @@
+	publishContext := make(map[string]string)
+
+	volID := pv.Spec.PersistentVolumeSource.CSI.VolumeHandle
+	if acquired := ns.VolumeLocks.TryAcquire(volID); !acquired {
+		util.ErrorLogMsg(util.VolumeOperationAlreadyExistsFmt, volID)
+		return status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volID)
+	}
+	defer ns.VolumeLocks.Release(volID)
+
+	stagingParentPath := pluginPath + "/kubernetes.io/csi/pv/" +
+		pv.Name + "/globalmount"
+
+	util.DefaultLog("sending nodeStageVolume for volID: %s, stagingPath: %s",
+		volID, stagingParentPath)
+
+	deviceSecret, err := getSecret(c,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Namespace,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Name)
+	if err != nil {
+		util.ErrorLogMsg("getSecret failed for volID: %s, err: %v", volID, err)
+		return err
+	}
+
+	volumeContext := pv.Spec.PersistentVolumeSource.CSI.VolumeAttributes
+	volumeContext["volumeHealerContext"] = "true"
+
+	req := &csi.NodeStageVolumeRequest{
+		VolumeId:          volID,
+		PublishContext:    publishContext,
+		StagingTargetPath: stagingParentPath,
+		VolumeCapability: &csi.VolumeCapability{
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: accessModeStrToInt(string(pv.Spec.AccessModes[0])),
+			},
+		},
+		Secrets:       deviceSecret,
+		VolumeContext: volumeContext,
+	}
+	if pv.Spec.PersistentVolumeSource.CSI.FSType == "block" {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Block{
+			Block: &csi.VolumeCapability_BlockVolume{},
+		}
+	} else {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Mount{
+			Mount: &csi.VolumeCapability_MountVolume{
+				FsType:     pv.Spec.PersistentVolumeSource.CSI.FSType,
+				MountFlags: pv.Spec.MountOptions,
+			},
+		}
+	}
+
+	_, err = ns.NodeStageVolume(context.TODO(), req)
+	if err != nil {
+		util.ErrorLogMsg("nodeStageVolume request failed, volID: %s, stagingPath: %s, err: %v",
+			volID, stagingParentPath, err)
+		return err
+	}
+
+	return nil
+}
+
+// RunVolumeHealer heal the volumes attached on a node.
+func RunVolumeHealer(ns *NodeServer, conf *util.Config) error {
+	c := util.NewK8sClient()
+	val, err := c.StorageV1().VolumeAttachments().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		util.ErrorLogMsg("list volumeAttachments failed, err: %v", err)
+		return err
+	}
+
+	var wg sync.WaitGroup
+	channel := make(chan error)
+	for i := range val.Items {
+		// skip if the volumeattachments doesn't belong to current node
+		if val.Items[i].Spec.NodeName != conf.NodeID {
+			continue
+		}
+		pvName := *val.Items[i].Spec.Source.PersistentVolumeName
+		pv, err := c.CoreV1().PersistentVolumes().Get(context.TODO(), pvName, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
+			util.ErrorLogMsg("get persistentVolumes failed for pv: %s, err: %v", pvName, err)
+			continue
+		}
+		// skip this volumeattachment if its pv is not bound
```

Comment:
Just to avoid corner cases, skip this if PV is marked for deletion.

Reply:
For all other cases, like "Released", "Pending", "Available" and "Failed", healing will be skipped.
i.e. Only if the PV state is in "Bound", the healer will run.

Hope I didn't miss anything here.

Reply:
If the PV is marked for deletion it will still be in a Bound state? just need to verify this one

Reply:
Is there any way to check this?
Because the state change will be too quick to observe in normal flow.

Reply:
Add:
TODO: check the pv delete annotations, like what happened when the pv is marked for delete.